include/nbl/video/IGPUFence.h

```diff
@@ -31,6 +33,94 @@ class IGPUFence : public core::IReferenceCounted, public IBackendObject
     virtual ~IGPUFence() = default;
 };
 
+class GPUEventWrapper : public core::Uncopyable
+{
+protected:
+    ILogicalDevice* mDevice;
+    core::smart_refctd_ptr<IGPUFence> mFence;
+public:
+    GPUEventWrapper(ILogicalDevice* dev, core::smart_refctd_ptr<IGPUFence>&& fence) : mDevice(dev), mFence(std::move(fence))
+    {
+    }
+    GPUEventWrapper(const GPUEventWrapper& other) = delete;
+    GPUEventWrapper(GPUEventWrapper&& other) noexcept : mFence(nullptr)
+    {
+        this->operator=(std::forward<GPUEventWrapper>(other));
+    }
+    virtual ~GPUEventWrapper()
+    {
+    }
+
+    GPUEventWrapper& operator=(const GPUEventWrapper& other) = delete;
+    inline GPUEventWrapper& operator=(GPUEventWrapper&& other) noexcept
+    {
+        mFence.operator=(std::move(other.mFence));
+        return *this;
+    }
+
+    template<class Clock=std::chrono::high_resolution_clock, class Duration=typename Clock::duration>
+    inline static std::chrono::time_point<Clock,Duration> default_wait()
+    {
+        return std::chrono::high_resolution_clock::now()+std::chrono::nanoseconds(50000ull); // 50 us
+    }
+
+    IGPUFence::E_STATUS waitFenceWrapper(IGPUFence* fence, uint64_t timeout);
+    IGPUFence::E_STATUS getFenceStatusWrapper(IGPUFence* fence);
+
+    template<class Clock=std::chrono::steady_clock, class Duration=typename Clock::duration>
+    inline bool wait_until(const std::chrono::time_point<Clock,Duration>& timeout_time)
+    {
+        auto currentClockTime = Clock::now();
+        do
+        {
+            uint64_t nanosecondsLeft = 0ull;
+            if (currentClockTime<timeout_time)
+                nanosecondsLeft = std::chrono::duration_cast<std::chrono::nanoseconds>(timeout_time-currentClockTime).count();
+            const IGPUFence::E_STATUS waitStatus = waitFenceWrapper(mFence.get(), nanosecondsLeft);
+            switch (waitStatus)
+            {
+            case IGPUFence::ES_ERROR:
+                return true;
+            case IGPUFence::ES_TIMEOUT:
+                break;
+            case IGPUFence::ES_SUCCESS:
+                return true;
+                break;
+            }
```

Comment:
maybe wrap the `switch` as well and return a bool, also make the wrapper protected at least?